scripts/gas-refund-program/transactions-validation/GRPBudgetGuardian.ts

```diff
@@ -34,49 +38,78 @@ export class GRPBudgetGuardian {
+      totalPSPRefundedForYear,
+      totalRefundedUSDByAddressForYear,
+      totalRefundedUSDByAddressForEpoch: {}, // no need to preload as validation runs on full epoch from scratch
     };
   }
 
-  totalRefundedAmountUSD(account: string) {
-    return this.state.totalRefundedAmountUSDByAddress[account] || ZERO_BN;
+  // ---------  PSP Global Yearly Budget Limit ---------
+  isMaxYearlyPSPGlobalBudgetSpent() {
+    return this.state.totalPSPRefundedForYear.isGreaterThanOrEqualTo(
+      MAX_PSP_GLOBAL_BUDGET_YEARLY,
+    );
   }
 
-  isMaxPSPGlobalBudgetSpent() {
-    return this.state.totalPSPRefunded.isGreaterThanOrEqualTo(
-      MAX_PSP_GLOBAL_BUDGET,
+  assertMaxYearlyPSPGlobalBudgetNotSpent() {
+    assert(
+      !this.isMaxYearlyPSPGlobalBudgetSpent(),
+      'Max PSP global budget spent',
     );
   }
 
-  isAccountUSDBudgetSpent(account: string) {
-    return this.totalRefundedAmountUSD(account).isGreaterThanOrEqualTo(
-      MAX_USD_ADDRESS_BUDGET,
+  increaseTotalRefundedPSP(amount: BigNumber) {
+    this.state.totalPSPRefundedForYear = (
+      this.state.totalPSPRefundedForYear || ZERO_BN
+    ).plus(amount);
+  }
+
+  // ---------  USD Per User Yearly Budget Limit ---------
+  totalYearlyRefundedUSD(account: string) {
+    return this.state.totalRefundedUSDByAddressForYear[account] || ZERO_BN;
+  }
+
+  hasSpentYearlyUSDBudget(account: string) {
+    return this.totalYearlyRefundedUSD(account).isGreaterThanOrEqualTo(
+      MAX_USD_ADDRESS_BUDGET_YEARLY,
     );
   }
 
-  assertMaxPSPGlobalBudgetNotReached() {
-    assert(!this.isMaxPSPGlobalBudgetSpent(), 'Max PSP global budget spent');
+  increaseYearlyRefundedUSD(account: string, usdAmount: BigNumber) {
+    this.state.totalRefundedUSDByAddressForYear[account] =
+      this.totalYearlyRefundedUSD(account).plus(usdAmount);
   }
 
-  increaseTotalAmountRefundedUSDForAccount(
-    account: string,
-    usdAmount: BigNumber | string,
-  ) {
-    this.state.totalRefundedAmountUSDByAddress[account] =
-      this.totalRefundedAmountUSD(account).plus(usdAmount);
+  // ---------  USD Per User Epoch Based Budget Limit ---------
+  totalRefundedUSDForEpoch(account: string) {
+    return this.state.totalRefundedUSDByAddressForEpoch[account] || ZERO_BN;
   }
 
-  increaseTotalPSPRefunded(amount: BigNumber | string) {
-    this.state.totalPSPRefunded = (this.state.totalPSPRefunded || ZERO_BN).plus(
-      amount,
+  hasSpentUSDBudgetForEpoch(account: string) {
+    return this.totalRefundedUSDForEpoch(account).isGreaterThanOrEqualTo(
+      MAX_USD_ADDRESS_BUDGET_EPOCH,
     );
   }
+
+  increaseRefundedUSDForEpoch(account: string, usdAmount: BigNumber) {
+    this.state.totalRefundedUSDByAddressForEpoch[account] =
+      this.totalRefundedUSDForEpoch(account).plus(usdAmount);
+  }
+
+  // ------------ cleaning -----
+  cleanYearlyBudgetState() {
+    this.state.totalPSPRefundedForYear = ZERO_BN;
+    this.state.totalRefundedUSDByAddressForYear = {};
+  }
+
+  cleanEpochBudgetState() {
```

Comment:
`resetEpochBudgetState` makes more sense to me.

Reply:
done 3e121195b8bc0cb3c27b6e31ea19f40d703c171a